pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/ReplicaGroupInstanceSelector.java

```diff
@@ -47,34 +51,108 @@
  * selected such that half the segments will come from S1 and other half from S2. If NUM_REPLICA_GROUPS_TO_QUERY value
  * is much greater than available servers, then ReplicaGroupInstanceSelector will behave similar to
  * BalancedInstanceSelector.
+ * <p>If AdaptiveServerSelection is enabled, a single snapshot of the server ranking is fetched. This ranking is
+ * referenced to pick the best available server for each segment. The algorithm ends up picking the minimum number of
+ * servers required to process a query because it references a single snapshot of the server rankings. Currently,
+ * NUM_REPLICA_GROUPS_TO_QUERY is not supported is AdaptiveServerSelection is enabled.
  */
 public class ReplicaGroupInstanceSelector extends BaseInstanceSelector {
 
-  public ReplicaGroupInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics) {
-    super(tableNameWithType, brokerMetrics);
+  public ReplicaGroupInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics,
+      AdaptiveServerSelector adaptiveServerSelector) {
+    super(tableNameWithType, brokerMetrics, adaptiveServerSelector);
   }
 
   @Override
   Map<String, String> select(List<String> segments, int requestId,
-      Map<String, List<String>> segmentToEnabledInstancesMap, Map<String, String> queryOptions) {
+      Map<String, List<String>> segmentToEnabledInstancesMap, Map<String, String> queryOptions,
+      @Nullable AdaptiveServerSelector adaptiveServerSelector) {
     Map<String, String> segmentToSelectedInstanceMap = new HashMap<>(HashUtil.getHashMapCapacity(segments.size()));
+
+    List<String> serverRankList = new ArrayList<>();
```

Comment:
This will create an extra list even when adaptive server selection is not enabled. Let's avoid that by wrapping the new logic under the if check.

Reply:
Done.